polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,281 @@
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument. Currently, all arguments
+/// with valid memref type determined by isValidMemRefType is a candidate
+/// argument, i.e., 'memref<?x!sycl.accessor>`.
+static bool isCandidateArg(Value arg) {
+  if (isValidMemRefType(arg.getType())) {
+    assert(isa<AccessorPtrType>(arg) &&
+           "Expecting valid memref type to be AccessorPtrType");
+    return true;
+  }
+  return false;
+}
+
+/// Populate \p candArgs with candidate arguments from \p call.
+static void
+collectCandidateArguments(CallOpInterface call,
+                          SmallVectorImpl<AccessorPtrType> &candArgs) {
+  assert(candArgs.empty() && "Expecting empty candArgs");
+  for (Value arg : call.getArgOperands())
+    if (isCandidateArg(arg))
+      candArgs.push_back(cast<AccessorPtrType>(arg));
+}
+
+/// Returns the cloned version of \p func.
+static FunctionOpInterface cloneFunction(FunctionOpInterface func) {
+  FunctionOpInterface clonedFunc = func.clone();
+  std::string newFnName = (func.getName() + ".specialized").str();
+#ifndef NDEBUG
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  module->walk([newFnName](FunctionOpInterface func) {
+    assert(func.getName() != newFnName &&
+           "Expecting new function name to be unique");
+  });
+#endif // NDEBUG
+  clonedFunc.setName(newFnName);
+  OpBuilder builder(func);
+  builder.insert(clonedFunc);
+  privatize(clonedFunc);
+  return clonedFunc;
+}
+
+/// Add attribute 'sycl.inner.disjoint' to arguments of \p func that satisfy \p
+/// predicate.
+static void setInnerDisjointAttribute(FunctionOpInterface func,
+                                      std::function<bool(Value)> predicate) {
+  constexpr StringRef innerDisjointAttrName = "sycl.inner.disjoint";
+  for (unsigned i = 0; i < func.getNumArguments(); ++i)
+    if (predicate(func.getArgument(i)))
+      func.setArgAttr(i, innerDisjointAttrName,
+                      UnitAttr::get(func->getContext()));
+}
+
+/// Update the callee of the call \p call to \p callee.
+static void updateCallee(CallOpInterface call, FunctionOpInterface callee) {
+  // TODO: find a generic way to update callee of a CallOpInterface.
+  assert(call->hasAttr("callee") &&
+         "Expecting the call to have callee attribute");
+  call->setAttr("callee",
+                SymbolRefAttr::get(callee.getContext(), callee.getName()));
```

Comment:
Instead of failing in this case, we could bail out and give up on the transformation. I'd remove the `assert` here and check that the call is in fact a `func.call` earlier in the pass, skipping other kind of call operations.

Reply:
I noticed that there are more than one `CallOpInterface` callee can be changed this way, e.g., `fir.call`. Do you think is worth allowing those non `func.call` `CallOpInterface` in this pass?

Reply:
If we have to limit to a certain operation (e,g, func::CallOp) I would give up on attempting to use a CallOpInterface all together.

Reply:
We may change the call from `func.call` to call operation from other dialect, e.g., `sycl.call`, to keep the pass more stable, I prefer not to give up on using `CallOpInterface`. Added method `setCalleeFromCallable` to `CallOpInterface`. WDYT?

Reply:
`setCalleeFromCallable` LGTM as long as we upstream that

Reply:
Will work on that after this pass is completed.